src/storage/src/source/antichain.rs

```diff
@@ -172,3 +209,205 @@ impl OffsetAntichain {
         }
     }
 }
+
+/// A wrapper around [`MutableAntichain`] that allows adding (inserting all
+/// contents with a `+1`) and subtracting (inserting all contents with a `-1`)
+/// of whole [`OffsetAntichains`](OffsetAntichain).
+///
+/// The frontier of this mutable antichain can be revealed in the form of an
+/// [`OffsetAntichain`].
+#[derive(Debug)]
+pub struct MutableOffsetAntichain {
+    inner: MutableAntichain<PartitionOffset>,
+}
+
+impl MutableOffsetAntichain {
+    /// Creates a new, empty [`MutableOffsetAntichain`].
+    pub fn new() -> Self {
+        Self {
+            inner: MutableAntichain::new(),
+        }
+    }
+
+    /// Inserts all partition/offset pairs contained in the given
+    /// [`OffsetAntichain`] into this [`MutableOffsetAntichain`], with a `diff`
+    /// of `+1`.
+    ///
+    /// In laymans terms, this adds the contained partition/offset pairs.
+    pub fn add(&mut self, offsets: &OffsetAntichain) {
+        let iter = offsets
+            .iter()
+            .map(|(pid, offset)| (PartitionOffset::new(pid.clone(), *offset), 1));
+        self.inner.update_iter(iter);
+    }
+
+    /// Inserts all partition/offset pairs contained in the given
+    /// [`OffsetAntichain`] into this [`MutableOffsetAntichain`], with a `diff`
+    /// of `-1`.
+    ///
+    /// In laymans terms, this subtracts the contained partition/offset pairs.
+    pub fn subtract(&mut self, offsets: &OffsetAntichain) {
+        let iter = offsets
+            .iter()
+            .map(|(pid, offset)| (PartitionOffset::new(pid.clone(), *offset), -1));
+        self.inner.update_iter(iter);
+    }
+
+    /// Reveals the minimal elements with positive count.
+    ///
+    /// In laymans terms, this returns an [`OffsetAntichain`] that contains all
+    /// partitions with positive counts, and their respective minimal offset.
+    pub fn frontier(&self) -> OffsetAntichain {
+        let mut result = OffsetAntichain::new();
+
+        for PartitionOffset { partition, offset } in self.inner.frontier().iter() {
+            result.insert(partition.clone(), *offset);
+        }
+
+        result
+    }
+}
+
+// NOTE: Ord is only required as an implementation detail of `MutableAntichain`,
+// but it feels iffy.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct PartitionOffset {
+    partition: PartitionId,
+    offset: MzOffset,
+}
+
+impl PartitionOffset {
+    fn new(partition: PartitionId, offset: MzOffset) -> Self {
+        PartitionOffset { partition, offset }
+    }
+}
+
+impl PartialOrder for PartitionOffset {
+    fn less_equal(&self, other: &Self) -> bool {
+        // Only offsets for the same partition are comparable!
+        if self.partition != other.partition {
+            false
+        } else {
+            self.offset <= other.offset
+        }
+    }
```

Comment:
this is a nice lil example of what a partial order actually IS

Reply:
TODO: manually implement `Ord` to show that it is compatible with `PartialOrder`